File: mobile/lib/services/reviews_service.dart
import '../config/api_client.dart';
import '../models/review.dart';

class ReviewsService {
  final ApiClient _apiClient = ApiClient();

  Future<ReviewListResponse> list({
    String? propertyId,
    String? revieweeId,
    int page = 1,
    int limit = 20,
  }) async {
    final response = await _apiClient.dio.get(
      '/reviews',
      queryParameters: {
        if (propertyId != null) 'propertyId': propertyId,
        if (revieweeId != null) 'revieweeId': revieweeId,
        'page': page,
        'limit': limit,
      },
    );
    return ReviewListResponse.fromJson(Map<String, dynamic>.from(response.data as Map));
  }

  Future<ReviewListResponse> listMine({String? propertyId, int page = 1, int limit = 20}) async {
    final response = await _apiClient.dio.get(
      '/reviews/mine',
      queryParameters: {
        if (propertyId != null) 'propertyId': propertyId,
        'page': page,
        'limit': limit,
      },
    );
    return ReviewListResponse.fromJson(Map<String, dynamic>.from(response.data as Map));
  }

  Future<Review> getById(String id) async {
    final response = await _apiClient.dio.get('/reviews/$id');
    return Review.fromJson(Map<String, dynamic>.from(response.data as Map));
  }

  Future<ReviewListResponse> getByProperty(String propertyId, {int page = 1, int limit = 20}) async {
    final response = await _apiClient.dio.get(
      '/reviews/property/$propertyId',
      queryParameters: {'page': page, 'limit': limit},
    );
    return ReviewListResponse.fromJson(Map<String, dynamic>.from(response.data as Map));
  }

  Future<ReviewListResponse> getBySeller(String revieweeId, {int page = 1, int limit = 20}) async {
    final response = await _apiClient.dio.get(
      '/reviews/seller/$revieweeId',
      queryParameters: {'page': page, 'limit': limit},
    );
    return ReviewListResponse.fromJson(Map<String, dynamic>.from(response.data as Map));
  }

  Future<Review> create(Map<String, dynamic> payload) async {
    final response = await _apiClient.dio.post('/reviews', data: payload);
    return Review.fromJson(Map<String, dynamic>.from(response.data as Map));
  }

  Future<Review> update(String id, Map<String, dynamic> payload) async {
    final response = await _apiClient.dio.patch('/reviews/$id', data: payload);
    return Review.fromJson(Map<String, dynamic>.from(response.data as Map));
  }

  Future<void> voteHelpful(String id, bool isHelpful) async {
    await _apiClient.dio.post('/reviews/$id/helpful', data: {'isHelpful': isHelpful});
  }

  Future<void> report(String id, String reason, {String? details}) async {
    await _apiClient.dio.post('/reviews/$id/report', data: {
      'reportReason': reason,
      if (details != null) 'reportDetails': details,
    });
  }

  Future<void> reply(String id, String replyText) async {
    await _apiClient.dio.post('/reviews/$id/reply', data: {'replyText': replyText});
  }
}

File: web/src/services/reviews.service.ts
import apiClient from '../config/api';
import type {
  Review,
  ReviewFilter,
  ReviewListResponse,
  CreateReviewPayload,
  UpdateReviewPayload,
  ReportReviewPayload,
  ReviewReply,
} from '../types/review';

export const reviewsService = {
  async list(filter: ReviewFilter = {}): Promise<ReviewListResponse> {
    const response = await apiClient.get('/reviews', { params: filter });
    return response.data;
  },

  async listMine(filter: ReviewFilter = {}): Promise<ReviewListResponse> {
    const response = await apiClient.get('/reviews/mine', { params: filter });
    return response.data;
  },

  async getByProperty(propertyId: string, filter: ReviewFilter = {}): Promise<ReviewListResponse> {
    const response = await apiClient.get(`/reviews/property/${propertyId}`, { params: filter });
    return response.data;
  },

  async getBySeller(revieweeId: string, filter: ReviewFilter = {}): Promise<ReviewListResponse> {
    const response = await apiClient.get(`/reviews/seller/${revieweeId}`, { params: filter });
    return response.data;
  },

  async getById(id: string): Promise<Review> {
    const response = await apiClient.get(`/reviews/${id}`);
    return response.data;
  },

  async create(payload: CreateReviewPayload): Promise<Review> {
    const response = await apiClient.post('/reviews', payload);
    return response.data;
  },

  async update(id: string, payload: UpdateReviewPayload): Promise<Review> {
    const response = await apiClient.patch(`/reviews/${id}`, payload);
    return response.data;
  },

  async voteHelpful(id: string, isHelpful: boolean): Promise<void> {
    await apiClient.post(`/reviews/${id}/helpful`, { isHelpful });
  },

  async report(id: string, payload: ReportReviewPayload): Promise<void> {
    await apiClient.post(`/reviews/${id}/report`, payload);
  },

  async reply(id: string, replyText: string): Promise<ReviewReply> {
    const response = await apiClient.post(`/reviews/${id}/reply`, { replyText });
    return response.data;
  },
};
